### PR TITLE
client: call .Endpoints() in dial() in client/v3/client.go instead of accessing cfg.Endpoints directly

### LINKED ISSUE #13201: accessing cfg.Endpoints in dial() in client/v3/client.go has race with SetEndpoints
https://github.com/etcd-io/etcd/blob/0cdd558361c6bdbbd9e4023558e2f6ece71c18ad/client/v3/client.go#L299 needs to be `c.Endpoints()` to avoid data race with `SetEndpoints`

### PR DESCRIPTION
**Pull request by yishuT**
https://github.com/etcd-io/etcd/blob/0cdd558361c6bdbbd9e4023558e2f6ece71c18ad/client/v3/client.go#L299 accesses
endpoints without acquiring lock. Fix it to call Endpoints()

Fix #13201

### CODE REVIEW

**Comment by serathius** [comment]
I agree that this variable needs a lock to prevent data race with `SetEndpoints`, but I would also not expect that `SetEndpoints` should modify `Config`. 
I would not expect config structs like `client.v3.Config` to be ever mutated by consumers. I expect that `SetEndpoints` was added later. Mutex withing Client doesn't directly specify which variables are protected by it.  

Instead of of just adding lock, maybe we should consider adding dedicated endpoint field to `Client` struct and specifying explicitly that is protected by mutex?

**Comment by yishuT** [comment]
yeah, I agree.  I only fixed what our race detector complained but didn't read the code in detail.  There is one thing that if `SetEndpoints` is called for a new list of endpoints after client is init, now this client has two different list of endpoints, one in client.cfg, the other in client.endpoint.

**Comment by yishuT** [comment]
nvm, we don't do deep copy in SetEndpoints

**Review by serathius — APPROVED** [review]
LG

**Comment by yishuT** [comment]
any updates, or should I do something?

**Review by spzala — APPROVED** [review]
lgtm. Can you please squash commits into one? Thanks @yishuT

**Comment by yishuT** [comment]
squash commits

**Comment by spzala** [comment]
> squash commits

Thanks! 
CI is good, also tested locally.

**Comment by twz123** [comment]
Now that this PR is available in etcd 3.6, there's some GRPC warnings being issued when creating a client via [client.New](https://pkg.go.dev/go.etcd.io/etcd/client/v3#New):

```text
2025/10/15 08:16:40 WARNING: [core] [Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "127.0.0.1:2379", ServerName: "127.0.0.1", }. Err: connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:2379: operation was canceled"
```

Is that expected? Everything else works as intended, `New` returns without error, and the returned client is usable. However, a "warning" implies that something is not as it should be... :thinking:

File: client/v3/client.go [hunk 1]
```diff
@@ -390,6 +392,8 @@ func newClient(cfg *Config) (*Client, error) {
 		client.cancel()
 		return nil, fmt.Errorf("at least one Endpoint is required in client config")
 	}
+	client.SetEndpoints(cfg.Endpoints...)
```
Comment (twz123) [inline client/v3/client.go:395]: This calls `client.resolver.SetEndpoints(eps)`, which hasn't been called before. I cannot really judge if this is a problem or not, but I observe GRPC warnings being logged with this:

```text
2025/10/15 12:55:29 WARNING: [core] [Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "127.0.0.1:2379", ServerName: "127.0.0.1:2379", }. Err: connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:2379: operation was canceled"
```
or 

```
2025/10/15 12:56:24 WARNING: [core] [Channel #1 SubChannel #2] grpc: addrConn.createTransport failed to connect to {Addr: "127.0.0.1:2379", ServerName: "127.0.0.1:2379", }. Err: connection error: desc = "transport: authentication handshake failed: context canceled"
```

Replacing this with the following makes them disappear again:

```suggestion
	client.epMu.Lock()
	client.endpoints = cfg.Endpoints
	client.epMu.Unlock()
```

**Pull request merged**